// .ignored/@coinbase/onchainkit/esm/wallet/components/ConnectWallet.js
import { ConnectButton } from '@rainbow-me/rainbowkit';
import { useMemo, Children, isValidElement, useCallback } from 'react';
import { useAccount, useConnect } from 'wagmi';
import { IdentityProvider } from '../../identity/components/IdentityProvider.js';
import { Spinner } from '../../internal/components/Spinner.js';
import { findComponent } from '../../internal/utils/findComponent.js';
import { cn, pressable, text, color } from '../../styles/theme.js';
import { ConnectButton as ConnectButton$1 } from './ConnectButton.js';
import { ConnectWalletText } from './ConnectWalletText.js';
import { useWalletContext } from './WalletProvider.js';
import { jsx } from 'react/jsx-runtime';
function ConnectWallet({
  children,
  className,
  // In a few version we will officially depracate this prop,
  // but for now we will keep it for backward compatibility.
  text: text$1 = 'Connect Wallet',
  withWalletAggregator = false
}) {
  // Core Hooks
  const _useWalletContext = useWalletContext(),
    isOpen = _useWalletContext.isOpen,
    setIsOpen = _useWalletContext.setIsOpen;
  const _useAccount = useAccount(),
    accountAddress = _useAccount.address,
    status = _useAccount.status;
  const _useConnect = useConnect(),
    connectors = _useConnect.connectors,
    connect = _useConnect.connect,
    connectStatus = _useConnect.status;

  // Get connectWalletText from children when present,
  // this is used to customize the connect wallet button text
  const _useMemo = useMemo(() => {
      const childrenArray = Children.toArray(children);
      return {
        connectWalletText: childrenArray.find(findComponent(ConnectWalletText))
      };
    }, [children]),
    connectWalletText = _useMemo.connectWalletText;

  // Remove connectWalletText from children if present
  const childrenWithoutConnectWalletText = useMemo(() => {
    return Children.map(children, child => {
      if ( /*#__PURE__*/isValidElement(child) && child.type === ConnectWalletText) {
        return null;
      }
      return child;
    });
  }, [children]);

  // Wallet connect status
  const connector = connectors[0];
  const isLoading = connectStatus === 'pending' || status === 'connecting';

  // Handles
  const handleToggle = useCallback(() => {
    setIsOpen(!isOpen);
  }, [isOpen, setIsOpen]);
  if (status === 'disconnected') {
    if (withWalletAggregator) {
      return /*#__PURE__*/jsx(ConnectButton.Custom, {
        children: ({
          openConnectModal
        }) => /*#__PURE__*/jsx("div", {
          className: "flex",
          "data-testid": "ockConnectWallet_Container",
          children: /*#__PURE__*/jsx(ConnectButton$1, {
            className: className,
            connectWalletText: connectWalletText,
            onClick: () => openConnectModal(),
            text: text$1
          })
        })
      });
    }
    return /*#__PURE__*/jsx("div", {
      className: "flex",
      "data-testid": "ockConnectWallet_Container",
      children: /*#__PURE__*/jsx(ConnectButton$1, {
        className: className,
        connectWalletText: connectWalletText,
        onClick: () => connect({
          connector
        }),
        text: text$1
      })
    });
  }
  if (isLoading) {
    return /*#__PURE__*/jsx("div", {
      className: "flex",
      "data-testid": "ockConnectWallet_Container",
      children: /*#__PURE__*/jsx("button", {
        type: "button",
        "data-testid": "ockConnectAccountButtonInner",
        className: cn(pressable.primary, text.headline, color.inverse, 'inline-flex min-w-[153px] items-center justify-center rounded-xl px-4 py-3', pressable.disabled, className),
        disabled: true,
        children: /*#__PURE__*/jsx(Spinner, {})
      })
    });
  }
  return /*#__PURE__*/jsx(IdentityProvider, {
    address: accountAddress,
    children: /*#__PURE__*/jsx("div", {
      className: "flex gap-4",
      "data-testid": "ockConnectWallet_Container",
      children: /*#__PURE__*/jsx("button", {
        type: "button",
        "data-testid": "ockConnectWallet_Connected",
        className: cn(pressable.secondary, 'rounded-xl px-4 py-3', isOpen && 'bg-ock-secondary-active hover:bg-ock-secondary-active', className),
        onClick: handleToggle,
        children: /*#__PURE__*/jsx("div", {
          className: "flex gap-2",
          children: childrenWithoutConnectWalletText
        })
      })
    })
  });
}
export { ConnectWallet };
