// .ignored/@coinbase/onchainkit/esm/identity/components/IdentityLayout.js
import { useMemo, Children } from 'react';
import { findComponent } from '../../internal/utils/findComponent.js';
import { cn, background, pressable, color } from '../../styles/theme.js';
import { usePopover } from '../hooks/usePopover.js';
import { Address } from './Address.js';
import { Avatar } from './Avatar.js';
import { EthBalance } from './EthBalance.js';
import { Name } from './Name.js';
import { jsxs, jsx } from 'react/jsx-runtime';
const noop = () => {};
function IdentityLayout({
  children,
  className,
  onClick
}) {
  const _useMemo = useMemo(() => {
      const childrenArray = Children.toArray(children);
      return {
        avatar: childrenArray.find(findComponent(Avatar)),
        name: childrenArray.find(findComponent(Name)),
        address: childrenArray.find(findComponent(Address)),
        ethBalance: childrenArray.find(findComponent(EthBalance))
      };
    }, [children]),
    avatar = _useMemo.avatar,
    name = _useMemo.name,
    address = _useMemo.address,
    ethBalance = _useMemo.ethBalance;
  const _usePopover = usePopover(onClick),
    handleClick = _usePopover.handleClick,
    handleMouseEnter = _usePopover.handleMouseEnter,
    handleMouseLeave = _usePopover.handleMouseLeave,
    showPopover = _usePopover.showPopover,
    popoverText = _usePopover.popoverText;
  return /*#__PURE__*/jsxs("div", {
    className: cn(background.default, 'flex items-center space-x-4 px-4 py-1', onClick && `${pressable.default} relative`, className),
    "data-testid": "ockIdentityLayout_container",
    onClick: handleClick,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onKeyUp: noop,
    onKeyDown: noop,
    children: [avatar, /*#__PURE__*/jsxs("div", {
      className: "flex flex-col",
      children: [name, address && !ethBalance && address, !address && ethBalance && ethBalance, address && ethBalance && /*#__PURE__*/jsxs("div", {
        className: "flex items-center gap-1",
        children: [address, /*#__PURE__*/jsx("span", {
          className: color.foregroundMuted,
          children: "\xB7"
        }), ethBalance]
      })]
    }), showPopover && /*#__PURE__*/jsxs("div", {
      className: cn(background.inverse, color.foreground, 'absolute top-[calc(100%_-_5px)] left-[46px] z-10 rounded px-2 py-1 shadow-[0px_4px_8px_rgba(0,0,0,0.1)]'),
      "data-testid": "ockIdentityLayout_copy",
      children: [popoverText, /*#__PURE__*/jsx("div", {
        className: cn('absolute top-[-5px] left-6 h-0 w-0', 'border-x-[5px] border-x-transparent border-b-[5px] border-b-[color:var(--bg-ock-inverse)] border-solid'),
        "data-testid": "ockIdentityLayout_copyArrow"
      })]
    })]
  });
}
export { IdentityLayout };
