// permissionless/_esm/actions/erc7579/accountId.js
import { ContractFunctionExecutionError, decodeFunctionResult, encodeFunctionData } from "viem";
import { parseAccount } from "../../utils/index.js";
import { AccountOrClientNotFoundError } from "../../utils/signUserOperationHashWithECDSA.js";
export async function accountId(client, args) {
    let account_ = client.account;
    if (args) {
        account_ = args.account;
    }
    if (!account_) {
        throw new AccountOrClientNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = parseAccount(account_);
    const publicClient = account.client;
    const abi = [
        {
            name: "accountId",
            type: "function",
            stateMutability: "view",
            inputs: [],
            outputs: [
                {
                    type: "string",
                    name: "accountImplementationId"
                }
            ]
        }
    ];
    try {
        return await publicClient.readContract({
            abi,
            functionName: "accountId",
            address: account.address
        });
    }
    catch (error) {
        if (error instanceof ContractFunctionExecutionError) {
            const factory = await account.getFactory();
            const factoryData = await account.getFactoryData();
            const result = await publicClient.call({
                factory: factory,
                factoryData: factoryData,
                to: account.address,
                data: encodeFunctionData({
                    abi,
                    functionName: "accountId"
                })
            });
            if (!result || !result.data) {
                throw new Error("accountId result is empty");
            }
            return decodeFunctionResult({
                abi,
                functionName: "accountId",
                data: result.data
            });
        }
        throw error;
    }
}
