// permissionless/_esm/actions/erc7579/installModules.js
import { encodeFunctionData, getAddress } from "viem";
import { getAction } from "viem/utils";
import { parseAccount } from "../../utils/index.js";
import { AccountOrClientNotFoundError } from "../../utils/signUserOperationHashWithECDSA.js";
import { sendUserOperation } from "../smartAccount/sendUserOperation.js";
import { parseModuleTypeId } from "./supportsModule.js";
export async function installModules(client, parameters) {
    const { account: account_ = client.account, maxFeePerGas, maxPriorityFeePerGas, nonce, middleware, modules } = parameters;
    if (!account_) {
        throw new AccountOrClientNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = parseAccount(account_);
    const installModulesCallData = await account.encodeCallData(await Promise.all(modules.map(({ type, address, context }) => ({
        to: account.address,
        value: BigInt(0),
        data: encodeFunctionData({
            abi: [
                {
                    name: "installModule",
                    type: "function",
                    stateMutability: "nonpayable",
                    inputs: [
                        {
                            type: "uint256",
                            name: "moduleTypeId"
                        },
                        {
                            type: "address",
                            name: "module"
                        },
                        {
                            type: "bytes",
                            name: "initData"
                        }
                    ],
                    outputs: []
                }
            ],
            functionName: "installModule",
            args: [
                parseModuleTypeId(type),
                getAddress(address),
                context
            ]
        })
    }))));
    return getAction(client, (sendUserOperation), "sendUserOperation")({
        userOperation: {
            sender: account.address,
            maxFeePerGas: maxFeePerGas,
            maxPriorityFeePerGas: maxPriorityFeePerGas,
            callData: installModulesCallData,
            nonce: nonce
        },
        account: account,
        middleware
    });
}
