// .ignored/@coinbase/onchainkit/esm/transaction/components/Transaction.js
import { cn } from '../../styles/theme.js';
import { useIsMounted } from '../../useIsMounted.js';
import { useOnchainKit } from '../../useOnchainKit.js';
import { TransactionProvider } from './TransactionProvider.js';
import { jsx } from 'react/jsx-runtime';
function Transaction({
  calls,
  capabilities,
  chainId,
  className,
  children,
  contracts,
  onError,
  onStatus,
  onSuccess
}) {
  const isMounted = useIsMounted();
  // prevents SSR hydration issue
  if (!isMounted) {
    return null;
  }
  const _useOnchainKit = useOnchainKit(),
    chain = _useOnchainKit.chain;
  // If chainId is not provided,
  // use the default chainId from the OnchainKit context
  const accountChainId = chainId ? chainId : chain.id;
  return /*#__PURE__*/jsx(TransactionProvider, {
    calls: calls,
    capabilities: capabilities,
    chainId: accountChainId,
    contracts: contracts,
    onError: onError,
    onStatus: onStatus,
    onSuccess: onSuccess,
    children: /*#__PURE__*/jsx("div", {
      className: cn(className, 'flex w-full flex-col gap-2'),
      children: children
    })
  });
}
export { Transaction };
