// permissionless@0.1.45/_cjs/actions/erc7579/installModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installModule = void 0;
const viem_1 = require("viem");
const utils_1 = require("viem/utils");
const utils_2 = require("../../utils/index.js");
const signUserOperationHashWithECDSA_1 = require("../../utils/signUserOperationHashWithECDSA.js");
const sendUserOperation_1 = require("../smartAccount/sendUserOperation.js");
const supportsModule_1 = require("./supportsModule.js");
async function installModule(client, parameters) {
    const { account: account_ = client.account, maxFeePerGas, maxPriorityFeePerGas, nonce, middleware, address, context } = parameters;
    if (!account_) {
        throw new signUserOperationHashWithECDSA_1.AccountOrClientNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = (0, utils_2.parseAccount)(account_);
    const installModuleCallData = await account.encodeCallData({
        to: account.address,
        value: BigInt(0),
        data: (0, viem_1.encodeFunctionData)({
            abi: [
                {
                    name: "installModule",
                    type: "function",
                    stateMutability: "nonpayable",
                    inputs: [
                        {
                            type: "uint256",
                            name: "moduleTypeId"
                        },
                        {
                            type: "address",
                            name: "module"
                        },
                        {
                            type: "bytes",
                            name: "initData"
                        }
                    ],
                    outputs: []
                }
            ],
            functionName: "installModule",
            args: [
                (0, supportsModule_1.parseModuleTypeId)(parameters.type),
                (0, viem_1.getAddress)(address),
                context
            ]
        })
    });
    return (0, utils_1.getAction)(client, (sendUserOperation_1.sendUserOperation), "sendUserOperation")({
        userOperation: {
            sender: account.address,
            maxFeePerGas: maxFeePerGas,
            maxPriorityFeePerGas: maxPriorityFeePerGas,
            callData: installModuleCallData,
            nonce: nonce
        },
        account: account,
        middleware
    });
}
exports.installModule = installModule;
