// permissionless@0.1.45/_cjs/actions/erc7579/isModuleInstalled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isModuleInstalled = void 0;
const viem_1 = require("viem");
const utils_1 = require("../../utils/index.js");
const signUserOperationHashWithECDSA_1 = require("../../utils/signUserOperationHashWithECDSA.js");
const supportsModule_1 = require("./supportsModule.js");
async function isModuleInstalled(client, parameters) {
    const { account: account_ = client.account, address, context } = parameters;
    if (!account_) {
        throw new signUserOperationHashWithECDSA_1.AccountOrClientNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = (0, utils_1.parseAccount)(account_);
    const publicClient = account.client;
    const abi = [
        {
            name: "isModuleInstalled",
            type: "function",
            stateMutability: "view",
            inputs: [
                {
                    type: "uint256",
                    name: "moduleTypeId"
                },
                {
                    type: "address",
                    name: "module"
                },
                {
                    type: "bytes",
                    name: "additionalContext"
                }
            ],
            outputs: [
                {
                    type: "bool"
                }
            ]
        }
    ];
    try {
        return await publicClient.readContract({
            abi,
            functionName: "isModuleInstalled",
            args: [
                (0, supportsModule_1.parseModuleTypeId)(parameters.type),
                (0, viem_1.getAddress)(address),
                context
            ],
            address: account.address
        });
    }
    catch (error) {
        if (error instanceof viem_1.ContractFunctionExecutionError) {
            const factory = await account.getFactory();
            const factoryData = await account.getFactoryData();
            const result = await publicClient.call({
                factory: factory,
                factoryData: factoryData,
                to: account.address,
                data: (0, viem_1.encodeFunctionData)({
                    abi,
                    functionName: "isModuleInstalled",
                    args: [
                        (0, supportsModule_1.parseModuleTypeId)(parameters.type),
                        (0, viem_1.getAddress)(address),
                        context
                    ]
                })
            });
            if (!result || !result.data) {
                throw new Error("accountId result is empty");
            }
            return (0, viem_1.decodeFunctionResult)({
                abi,
                functionName: "isModuleInstalled",
                data: result.data
            });
        }
        throw error;
    }
}
exports.isModuleInstalled = isModuleInstalled;
