// permissionless@0.1.45/_cjs/utils/encode7579CallData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encode7579CallData = void 0;
const viem_1 = require("viem");
const supportsExecutionMode_1 = require("../actions/erc7579/supportsExecutionMode.js");
function encode7579CallData({ mode, callData }) {
    if (Array.isArray(callData) && mode?.type !== "batchcall") {
        throw new Error(`mode ${JSON.stringify(mode)} does not supported for batchcall calldata`);
    }
    const executeAbi = [
        {
            type: "function",
            name: "execute",
            inputs: [
                {
                    name: "execMode",
                    type: "bytes32",
                    internalType: "ExecMode"
                },
                {
                    name: "executionCalldata",
                    type: "bytes",
                    internalType: "bytes"
                }
            ],
            outputs: [],
            stateMutability: "payable"
        }
    ];
    if (Array.isArray(callData)) {
        return (0, viem_1.encodeFunctionData)({
            abi: executeAbi,
            functionName: "execute",
            args: [
                (0, supportsExecutionMode_1.encodeExecutionMode)(mode),
                (0, viem_1.encodeAbiParameters)([
                    {
                        name: "executionBatch",
                        type: "tuple[]",
                        components: [
                            {
                                name: "target",
                                type: "address"
                            },
                            {
                                name: "value",
                                type: "uint256"
                            },
                            {
                                name: "callData",
                                type: "bytes"
                            }
                        ]
                    }
                ], [
                    callData.map((arg) => {
                        return {
                            target: arg.to,
                            value: arg.value,
                            callData: arg.data
                        };
                    })
                ])
            ]
        });
    }
    return (0, viem_1.encodeFunctionData)({
        abi: executeAbi,
        functionName: "execute",
        args: [
            (0, supportsExecutionMode_1.encodeExecutionMode)(mode),
            (0, viem_1.concatHex)([
                callData.to,
                (0, viem_1.toHex)(callData.value, { size: 32 }),
                callData.data
            ])
        ]
    });
}
exports.encode7579CallData = encode7579CallData;
