// permissionless/_cjs/actions/erc7579.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.erc7579Actions = exports.uninstallModules = exports.uninstallModule = exports.supportsModule = exports.supportsExecutionMode = exports.isModuleInstalled = exports.installModules = exports.installModule = exports.accountId = void 0;
const accountId_1 = require("./erc7579/accountId.js");
Object.defineProperty(exports, "accountId", { enumerable: true, get: function () { return accountId_1.accountId; } });
const installModule_1 = require("./erc7579/installModule.js");
Object.defineProperty(exports, "installModule", { enumerable: true, get: function () { return installModule_1.installModule; } });
const installModules_1 = require("./erc7579/installModules.js");
Object.defineProperty(exports, "installModules", { enumerable: true, get: function () { return installModules_1.installModules; } });
const isModuleInstalled_1 = require("./erc7579/isModuleInstalled.js");
Object.defineProperty(exports, "isModuleInstalled", { enumerable: true, get: function () { return isModuleInstalled_1.isModuleInstalled; } });
const supportsExecutionMode_1 = require("./erc7579/supportsExecutionMode.js");
Object.defineProperty(exports, "supportsExecutionMode", { enumerable: true, get: function () { return supportsExecutionMode_1.supportsExecutionMode; } });
const supportsModule_1 = require("./erc7579/supportsModule.js");
Object.defineProperty(exports, "supportsModule", { enumerable: true, get: function () { return supportsModule_1.supportsModule; } });
const uninstallModule_1 = require("./erc7579/uninstallModule.js");
Object.defineProperty(exports, "uninstallModule", { enumerable: true, get: function () { return uninstallModule_1.uninstallModule; } });
const uninstallModules_1 = require("./erc7579/uninstallModules.js");
Object.defineProperty(exports, "uninstallModules", { enumerable: true, get: function () { return uninstallModules_1.uninstallModules; } });
function erc7579Actions(_args) {
    return (client) => ({
        accountId: (args) => (0, accountId_1.accountId)(client, args),
        installModule: (args) => (0, installModule_1.installModule)(client, args),
        installModules: (args) => (0, installModules_1.installModules)(client, args),
        isModuleInstalled: (args) => (0, isModuleInstalled_1.isModuleInstalled)(client, args),
        supportsExecutionMode: (args) => (0, supportsExecutionMode_1.supportsExecutionMode)(client, args),
        supportsModule: (args) => (0, supportsModule_1.supportsModule)(client, args),
        uninstallModule: (args) => (0, uninstallModule_1.uninstallModule)(client, args),
        uninstallModules: (args) => (0, uninstallModules_1.uninstallModules)(client, args)
    });
}
exports.erc7579Actions = erc7579Actions;
