// .ignored/@coinbase/onchainkit/esm/swap/utils/sendSwapTransactions.js
import { Capabilities } from '../../constants.js';
import { getRPCUrl } from '../../network/getRPCUrl.js';
import { sendSingleTransactions } from './sendSingleTransactions.js';
async function sendSwapTransactions({
  config,
  isSponsored,
  sendCallsAsync,
  sendTransactionAsync,
  updateLifecycleStatus,
  walletCapabilities,
  transactions
}) {
  if (walletCapabilities[Capabilities.AtomicBatch]?.supported) {
    // For batched transactions, we'll use `SwapProvider` to listen for calls to emit the `success` state
    updateLifecycleStatus({
      statusName: 'transactionPending'
    });
    const callsId = await sendCallsAsync({
      calls: transactions.map(({
        transaction
      }) => transaction),
      capabilities: isSponsored ? {
        paymasterService: {
          url: getRPCUrl()
        }
      } : {}
    });
    updateLifecycleStatus({
      statusName: 'transactionApproved',
      statusData: {
        callsId,
        transactionType: 'Batched'
      }
    });
  } else {
    await sendSingleTransactions({
      config,
      sendTransactionAsync,
      transactions,
      updateLifecycleStatus
    });
  }
}
export { sendSwapTransactions };
