// @coinbase/onchainkit@0.33.3/esm/OnchainKitProvider.js
import { createContext, useMemo } from 'react';
import { ONCHAIN_KIT_CONFIG, setOnchainKitConfig } from './OnchainKitConfig.js';
import { checkHashLength } from './internal/utils/checkHashLength.js';
import { jsx } from 'react/jsx-runtime';
const OnchainKitContext = /*#__PURE__*/createContext(ONCHAIN_KIT_CONFIG);

/**
 * Provides the OnchainKit React Context to the app.
 */
function OnchainKitProvider({
  address,
  apiKey,
  chain,
  children,
  projectId,
  rpcUrl,
  schemaId
}) {
  if (schemaId && !checkHashLength(schemaId, 64)) {
    throw Error('EAS schemaId must be 64 characters prefixed with "0x"');
  }
  const value = useMemo(() => {
    const onchainKitConfig = {
      address: address ?? null,
      apiKey: apiKey ?? null,
      chain: chain,
      projectId: projectId ?? null,
      rpcUrl: rpcUrl ?? null,
      schemaId: schemaId ?? null
    };
    setOnchainKitConfig(onchainKitConfig);
    return onchainKitConfig;
  }, [address, apiKey, chain, projectId, rpcUrl, schemaId]);
  return /*#__PURE__*/jsx(OnchainKitContext.Provider, {
    value: value,
    children: children
  });
}
export { OnchainKitContext, OnchainKitProvider };
